File: src/aai_agent/voice_cleaner.py
"""Text normalization for TTS output.

Cleans LLM-generated text so it sounds natural when spoken by a TTS system.
Handles numbers, currency, markdown artifacts, URLs, and more.
"""

from __future__ import annotations

import re

import inflect

# Pre-compiled regex patterns
_RE_CODE_BLOCKS = re.compile(r"```[\s\S]*?```")
_RE_INDENTED_CODE = re.compile(r"^(?:    |\t).+$", re.MULTILINE)
_RE_INLINE_CODE = re.compile(r"`([^`]+)`")
_RE_BOLD_ITALIC_STAR = re.compile(r"\*{1,3}([^*]+)\*{1,3}")
_RE_BOLD_ITALIC_UNDER = re.compile(r"_{1,3}([^_]+)_{1,3}")
_RE_HEADERS = re.compile(r"^#{1,6}\s+", re.MULTILINE)
_RE_LINKS = re.compile(r"\[([^\]]+)\]\([^)]*\)")
_RE_IMAGES = re.compile(r"!\[[^\]]*\]\([^)]*\)")
_RE_BULLETS = re.compile(r"^\s*[-*+]\s+", re.MULTILINE)
_RE_NUMBERED = re.compile(r"^\s*\d+\.\s+", re.MULTILINE)
_RE_BLOCKQUOTES = re.compile(r"^\s*>\s?", re.MULTILINE)
_RE_HORIZ_RULES = re.compile(r"^[-*_]{3,}\s*$", re.MULTILINE)
_RE_URLS = re.compile(r"https?://\S+")
_RE_CURRENCY = re.compile(r"([$£€¥])(\d+(?:,\d{3})*(?:\.\d{2})?)")
_RE_PERCENTAGES = re.compile(r"(\d+(?:\.\d+)?)%")
_RE_PHONE = re.compile(r"(\d{3})-(\d{3})-(\d{4})")
_RE_ORDINALS = re.compile(r"\b(\d{1,2})(st|nd|rd|th)\b")
_RE_NUMBERS = re.compile(r"(?<![:\w])\b\d+(?:\.\d+)?\b(?![:\w])")
_RE_SPACES = re.compile(r"[ \t]+")
_RE_NEWLINES = re.compile(r"\n{2,}")

_CURRENCY_MAP = {"$": "dollars", "£": "pounds", "€": "euros", "¥": "yen"}

_UNIT_MAP = {
    "Hz": "hertz",
    "kHz": "kilohertz",
    "MHz": "megahertz",
    "GHz": "gigahertz",
    "KB": "kilobytes",
    "MB": "megabytes",
    "GB": "gigabytes",
    "TB": "terabytes",
    "ms": "milliseconds",
    "kb": "kilobits",
    "Mb": "megabits",
    "Gb": "gigabits",
}
_UNIT_PATTERNS = {
    abbr: re.compile(rf"(\d)\s*{re.escape(abbr)}\b") for abbr in _UNIT_MAP
}

_p = inflect.engine()


def _num2words(n: int | float) -> str:
    return _p.number_to_words(n)  # type: ignore[return-value, arg-type]


def _ordinal(word: str) -> str:
    return _p.ordinal(word)  # type: ignore[return-value, arg-type]


def normalize_voice_text(text: str) -> str:
    """Run all normalization passes on *text* and return cleaned result."""
    # Strip markdown
    text = _RE_CODE_BLOCKS.sub("", text)
    text = _RE_INDENTED_CODE.sub("", text)
    text = _RE_INLINE_CODE.sub(r"\1", text)
    text = _RE_BOLD_ITALIC_STAR.sub(r"\1", text)
    text = _RE_BOLD_ITALIC_UNDER.sub(r"\1", text)
    text = _RE_HEADERS.sub("", text)
    text = _RE_LINKS.sub(r"\1", text)
    text = _RE_IMAGES.sub("", text)
    text = _RE_BULLETS.sub("", text)
    text = _RE_NUMBERED.sub("", text)
    text = _RE_BLOCKQUOTES.sub("", text)
    text = _RE_HORIZ_RULES.sub("", text)

    # Remove URLs
    text = _RE_URLS.sub("", text)

    # Currency: $7.95 → seven dollars and ninety-five cents
    def _expand_currency(m: re.Match) -> str:
        symbol, num = m.groups()
        num_clean = num.replace(",", "")
        currency = _CURRENCY_MAP.get(symbol, "currency")
        if "." in num_clean:
            dollars, cents = num_clean.split(".")
            return f"{_num2words(int(dollars))} {currency} and {_num2words(int(cents))} cents"
        return f"{_num2words(int(num_clean))} {currency}"

    text = _RE_CURRENCY.sub(_expand_currency, text)

    # Percentages: 100% → 100 percent
    text = _RE_PERCENTAGES.sub(r"\1 percent", text)

    # Phone numbers: 555-772-9140 → five five five, seven seven two, ...
    def _expand_phone(m: re.Match) -> str:
        return ", ".join(
            " ".join(_num2words(int(d)) for d in group) for group in m.groups()
        )

    text = _RE_PHONE.sub(_expand_phone, text)

    # Ordinals: 1st → first, 2nd → second
    text = _RE_ORDINALS.sub(lambda m: _ordinal(_num2words(int(m.group(1)))), text)

    # Units: Hz → hertz, kHz → kilohertz, etc.
    for abbr, full in _UNIT_MAP.items():
        text = _UNIT_PATTERNS[abbr].sub(rf"\1 {full}", text)

    # Numbers: 123 → one hundred and twenty-three
    def _expand_number(m: re.Match) -> str:
        raw = m.group(0)
        try:
            return _num2words(float(raw)) if "." in raw else _num2words(int(raw))
        except (ValueError, OverflowError):
            return raw

    text = _RE_NUMBERS.sub(_expand_number, text)

    # Symbols
    text = text.replace("°F", " degrees Fahrenheit")
    text = text.replace("°C", " degrees Celsius")
    text = text.replace("°", " degrees")
    text = text.replace("&", " and ")
    text = text.replace("+", " plus ")
    text = text.replace("→", "")
    text = text.replace("—", ", ")
    text = text.replace("–", ", ")

    # Collapse whitespace
    text = _RE_SPACES.sub(" ", text)
    text = _RE_NEWLINES.sub("\n", text)
    return text.strip()


class VoiceCleaner:
    """Backward-compatible wrapper around :func:`normalize_voice_text`."""

    def normalize(self, text: str) -> str:
        return normalize_voice_text(text)

File: platform/src/voice-cleaner.ts
// voice-cleaner.ts — Text normalization for TTS.
// Port of the Python voice_cleaner.py.

import { toWords, toWordsOrdinal } from "number-to-words";

// Pre-compiled regex patterns
const RE_CODE_BLOCKS = /```[\s\S]*?```/g;
const RE_INDENTED_CODE = /^(?:    |\t).+$/gm;
const RE_INLINE_CODE = /`([^`]+)`/g;
const RE_BOLD_ITALIC_STAR = /\*{1,3}([^*]+)\*{1,3}/g;
const RE_BOLD_ITALIC_UNDER = /_{1,3}([^_]+)_{1,3}/g;
const RE_HEADERS = /^#{1,6}\s+/gm;
const RE_LINKS = /\[([^\]]+)\]\([^)]*\)/g;
const RE_IMAGES = /!\[[^\]]*\]\([^)]*\)/g;
const RE_BULLETS = /^\s*[-*+]\s+/gm;
const RE_NUMBERED = /^\s*\d+\.\s+/gm;
const RE_BLOCKQUOTES = /^\s*>\s?/gm;
const RE_HORIZ_RULES = /^[-*_]{3,}\s*$/gm;
const RE_URLS = /https?:\/\/\S+/g;
const RE_CURRENCY = /([$£€¥])(\d+(?:,\d{3})*(?:\.\d{2})?)/g;
const RE_PERCENTAGES = /(\d+(?:\.\d+)?)%/g;
const RE_PHONE = /(\d{3})-(\d{3})-(\d{4})/g;
const RE_ORDINALS = /\b(\d{1,2})(st|nd|rd|th)\b/g;
const RE_NUMBERS = /(?<![:\w])\b\d+(?:\.\d+)?\b(?![:\w])/g;
const RE_SPACES = /[ \t]+/g;
const RE_NEWLINES = /\n{2,}/g;

const CURRENCY_MAP: Record<string, string> = {
  $: "dollars",
  "£": "pounds",
  "€": "euros",
  "¥": "yen",
};

const UNIT_MAP: Record<string, string> = {
  Hz: "hertz",
  kHz: "kilohertz",
  MHz: "megahertz",
  GHz: "gigahertz",
  KB: "kilobytes",
  MB: "megabytes",
  GB: "gigabytes",
  TB: "terabytes",
  ms: "milliseconds",
  kb: "kilobits",
  Mb: "megabits",
  Gb: "gigabits",
};

function num2words(n: number): string {
  if (Number.isInteger(n)) {
    return toWords(n);
  }
  // number-to-words doesn't support floats — split on decimal point
  const [intPart, decPart] = String(n).split(".");
  const intWords = toWords(parseInt(intPart));
  const decWords = decPart
    .split("")
    .map((d) => toWords(parseInt(d)))
    .join(" ");
  return `${intWords} point ${decWords}`;
}

/**
 * Normalize text for TTS output.
 * Strips markdown, expands numbers/currency/units, collapses whitespace.
 */
export function normalizeVoiceText(text: string): string {
  // Strip markdown
  text = text.replace(RE_CODE_BLOCKS, "");
  text = text.replace(RE_INDENTED_CODE, "");
  text = text.replace(RE_INLINE_CODE, "$1");
  text = text.replace(RE_BOLD_ITALIC_STAR, "$1");
  text = text.replace(RE_BOLD_ITALIC_UNDER, "$1");
  text = text.replace(RE_HEADERS, "");
  text = text.replace(RE_IMAGES, "");   // Must run before LINKS
  text = text.replace(RE_LINKS, "$1");
  text = text.replace(RE_BULLETS, "");
  text = text.replace(RE_NUMBERED, "");
  text = text.replace(RE_BLOCKQUOTES, "");
  text = text.replace(RE_HORIZ_RULES, "");

  // Remove URLs
  text = text.replace(RE_URLS, "");

  // Currency: $7.95 → seven dollars and ninety-five cents
  text = text.replace(RE_CURRENCY, (_match, symbol: string, num: string) => {
    const numClean = num.replace(/,/g, "");
    const currency = CURRENCY_MAP[symbol] ?? "currency";
    if (numClean.includes(".")) {
      const [dollars, cents] = numClean.split(".");
      return `${num2words(parseInt(dollars))} ${currency} and ${num2words(parseInt(cents))} cents`;
    }
    return `${num2words(parseInt(numClean))} ${currency}`;
  });

  // Percentages: 100% → 100 percent
  text = text.replace(RE_PERCENTAGES, "$1 percent");

  // Phone numbers: 555-772-9140 → five five five, seven seven two, ...
  text = text.replace(RE_PHONE, (_match, a: string, b: string, c: string) => {
    const expand = (group: string) =>
      group
        .split("")
        .map((d) => num2words(parseInt(d)))
        .join(" ");
    return `${expand(a)}, ${expand(b)}, ${expand(c)}`;
  });

  // Ordinals: 1st → first, 2nd → second
  text = text.replace(RE_ORDINALS, (_match, num: string) => {
    return toWordsOrdinal(parseInt(num));
  });

  // Units: Hz → hertz, kHz → kilohertz, etc.
  for (const [abbr, full] of Object.entries(UNIT_MAP)) {
    const re = new RegExp(`(\\d)\\s*${abbr.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}\\b`, "g");
    text = text.replace(re, `$1 ${full}`);
  }

  // Numbers: 123 → one hundred and twenty-three
  text = text.replace(RE_NUMBERS, (match) => {
    try {
      const n = match.includes(".") ? parseFloat(match) : parseInt(match);
      return num2words(n);
    } catch {
      return match;
    }
  });

  // Symbols
  text = text.replace(/°F/g, " degrees Fahrenheit");
  text = text.replace(/°C/g, " degrees Celsius");
  text = text.replace(/°/g, " degrees");
  text = text.replace(/&/g, " and ");
  text = text.replace(/\+/g, " plus ");
  text = text.replace(/→/g, "");
  text = text.replace(/—/g, ", ");
  text = text.replace(/–/g, ", ");

  // Collapse whitespace
  text = text.replace(RE_SPACES, " ");
  text = text.replace(RE_NEWLINES, "\n");
  return text.trim();
}
